### PR TITLE
refactor(server): refactor package sync workers

### PR DESCRIPTION
**Pull request by cschmatzler**

### CODE REVIEW

**Comment by cschmatzler** [comment]
**Note: This is an AI-generated code review by Claude 4.1 Opus (Anthropic)**

## Code Review for refactor(server): refactor package sync workers

### Summary
This refactoring consolidates the package synchronization logic into a more modular and maintainable architecture. The changes replace two separate workers (`UpdatePackagesWorker` and `UpdatePackageReleasesWorker`) with a unified sync worker and a dedicated release creation worker.

### ✅ Positive Changes

1. **Improved Architecture**
   - Consolidates `UpdatePackagesWorker` and `UpdatePackageReleasesWorker` into a single `SyncPackagesWorker`
   - Introduces `CreatePackageReleaseWorker` for atomic release creation
   - Better separation of concerns with clear responsibilities for each worker

2. **Enhanced Performance**
   - Individual release creation jobs allow for better parallelization and retry strategies
   - Unique job constraints prevent duplicate work (using scope, name, version as keys)
   - More efficient error handling with isolated failures

3. **Better Observability**
   - Clear logging at each stage of the sync process
   - Appsignal integration for monitoring individual package and release operations
   - Easier to track and debug issues with specific packages or releases

4. **Code Organization**
   - Cleaner separation between discovering packages, finding missing releases, and creating releases
   - More testable code with focused unit tests for each worker
   - Reduced coupling between different synchronization stages

### 🔍 Observations & Suggestions

1. **Resource Management**
   - The `limit` parameter in `find_missing_releases_for_existing_packages` (default 350) might need tuning based on API rate limits and job processing capacity
   - Consider adding configuration for the max number of concurrent `CreatePackageReleaseWorker` jobs to avoid overwhelming the system

2. **Error Handling**
   - Good use of max_attempts (3 for CreatePackageReleaseWorker, 10 for SyncPackagesWorker)
   - The error rescue in `CreatePackageReleaseWorker` logs but might benefit from more specific error types for different failure scenarios

3. **Testing Coverage**
   - Excellent test coverage for `CreatePackageReleaseWorker` including edge cases (pre-release versions, v-prefixed versions, multiple manifests)
   - Good consolidation of test cases from the old workers into focused test files

4. **Minor Improvements**
   - Consider extracting the hardcoded limit of 100 for package removal and creation into a configuration constant
   - The `@unsupported_packages` list could be moved to configuration or database for easier management

### 📊 Impact Analysis

- **Lines changed**: -461 net reduction (980 additions, 1441 deletions)
- **Test coverage**: Maintained with focused test suites
- **Runtime impact**: Should improve with better parallelization and retry logic
- **Maintenance**: Significantly improved with clearer separation of concerns

### 🔬 Test Coverage Deep Dive - VERIFIED ✅

After thorough analysis of the test changes and actual test implementations:

**Test Migration Status:**
- **5 test cases removed** from old workers (`UpdatePackagesWorker`, `UpdatePackageReleasesWorker`)
- **16 test cases added** in new workers (`CreatePackageReleaseWorker`, `SyncPackagesWorker`)
- **Net gain: +11 test cases**

**Coverage Verification - ALL EDGE CASES PRESERVED ✅:**

I can confirm that ALL the complex edge cases mentioned in my initial review have been successfully migrated and are fully tested:

1. **String interpolation in Package URLs** ✅ - Test exists: "handles Package.swift with String interpolation in package URLs"
2. **Header search paths** ✅ - Test exists: "handles Package.swift with header search paths"
3. **.byName references** ✅ - Test exists: "handles Package.swift with .byName product references"
4. **Manifest fixing logic** ✅ - Test exists: "verifies manifest fixing logic transforms package references correctly"
5. **Binary package references** ✅ - Test exists: "creates package release when Package.swift includes binary references"
6. **Multiple Swift manifests** ✅ - Test exists: "creates package release with multiple Package.swift manifests for different Swift versions"
7. **Pre-release versions** ✅ - Test exists: "creates package release with pre-release version"
8. **v-prefixed versions** ✅ - Test exists: "creates package release with v-prefixed version"
9. **Missing Package.swift** ✅ - Test exists: "skips Package.swift upload when none exists in repository"

**Additional Coverage Improvements:**
- The new test suite includes comprehensive testing of the manifest transformation logic with detailed assertions
- Tests now verify the actual transformed content, not just successful execution
- Error handling paths are better tested with specific error scenarios

**Manifest Transformation Logic:**
The critical manifest fixing logic remains intact in `server/lib/tuist/registry/swift/packages.ex` and continues to handle:
- `.byName` reference transformations
- String dependency resolution
- Package name normalization
- Custom package name handling

### ✅ Verdict

This is an excellent refactoring that successfully:
1. Improves code maintainability and performance
2. **Maintains 100% test coverage of all edge cases**
3. Actually **improves test coverage** with more comprehensive assertions
4. Preserves all critical manifest transformation logic

The concerns about lost test coverage have been thoroughly investigated and **all edge cases are properly covered** in the new test suite.

**Approved - Test coverage fully maintained and improved** 👍

File: server/lib/tuist/registry/swift/workers/sync_packages_worker.ex [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    update_releases = Map.get(args, :update_releases, true)
+
+    token_packages = Environment.github_token_update_packages()
+    token_releases = Environment.github_token_update_package_releases()
+
+    missing_versions_from_packages =
+      if update_packages, do: sync_packages_from_spi(token_packages, token_releases), else: []
+
+    missing_versions_from_releases =
+      if update_releases, do: find_missing_releases_for_existing_packages(limit, token_releases), else: []
```
Comment (fortmarek) [inline server/lib/tuist/registry/swift/workers/sync_packages_worker.ex]:
```suggestion
    missing_versions_from_new_packages =
      if update_packages, do: sync_packages_from_spi(token_packages, token_releases), else: []

    missing_versions_from_existing_packages =
      if update_releases, do: find_missing_releases_for_existing_packages(limit, token_releases), else: []
```

I find these names better describe what those variables are about, it took me a second before parsing that based on the current names

**Review by fortmarek — APPROVED** [review]
Thanks! Changes look good, just one minor comment 🙂
